blog: post for Build-Your-Own-Builder on GHA (#929)
slsa-framework/slsa: docs/_posts/2023-08-07-bring-your-own-builder-github.md

---
title: "Build your own SLSA 3+ provenance builder on GitHub Actions"

> author: "Andres Almiray (JReleaser), Adam Korczynski (AdaLogics), Philip Harrison (GitHub), Laurent Simon (Google)"

Reviewer: `AdaLogics` --> `Ada Logics`

## SLSA Java builders for JReleaser, Maven and Gradle

To validate the design of the BYOB framework and demonstrate its flexibility, we have partnered with new contributors to create three new builders for the [Java ecosystems](https://github.com/slsa-framework/slsa-github-generator/blob/main/README.md#builder-creation):

1.  A [JReleaser](https://github.com/jreleaser/release-action/tree/java#slsa-builder) Java builder which wraps the existing [GitHub Action for JReleaser](https://github.com/jreleaser/release-action) into a SLSA3-compliant builder. The integration was done by [aalmiray@](https://github.com/aalmiray), the maintainer of the [JReleaser](https://jreleaser.org) project. The resulting builder is hosted in the same repository as its original [Action](https://github.com/jreleaser/release-action/blob/java/.github/workflows/builder_slsa3.yml), so that JReleaser users can continue using the repositories they are already accustomed to.

> 2.  A [Maven builder](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/maven#readme), contributed by [AdaLogics](https://adalogics.com). This builder is currently hosted in the OpenSSF SLSA repository.

Reviewer: `AdaLogics` --> `Ada Logics`  -- not in the url though :)

> 3.  A [Gradle builder](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/gradle#readme), contributed by [AdaLogics](https://adalogics.com). This builder is currently hosted in the OpenSSF SLSA repository.

Reviewer: `AdaLogics` --> `Ada Logics` -- not in the url though :)

SLSA now provides and maintains official builders for [Go](2022/06/slsa-github-workflows), [Node.js](2023/05/bringing-improved-supply-chain-security-to-the-nodejs-ecosystem) and [Container](2023/06/slsa-github-worfklows-container-based) based projects.
But what if you don't use any of these languages or use custom tooling that isn't supported by the official builders?

To empower the community to create their own provenance builders and leverage the secure architecture of the official SLSA builders we are releasing the ["Build Your Own Builder" framework](https://github.com/slsa-framework/slsa-github-generator/tree/main#build-your-own-builder) for GitHub Actions. This makes it easy to take an existing GitHub Action (e.g. [JReleaser](https://jreleaser.org/)) and make it produce [SLSA Build Level 3 provenance](https://slsa.dev/spec/v1.0/requirements#provenance-generation).

> Writing a builder from scratch is a tedious multi-month effort. The BYOB framework streamlines this process and cuts the development time down to a few days. As a builder writer, you need not worry about keeping signing keys secure, isolation between builds, the creation of attestations; all this is handled seamlessly by the framework, using the [same security design principles](https://github.com/slsa-framework/slsa-github-generator/tree/main#specifications) as our existing builders.

Reviewer 2: nit style suggestion
```suggestion
Writing a builder from scratch is a tedious multi-month effort. The BYOB framework streamlines this process and cuts the development time down to a few days. As a builder author, you don't need to worry about keeping signing keys secure, isolation between builds, the creation of attestations; all this is handled seamlessly by the framework, using the [same security design principles](https://github.com/slsa-framework/slsa-github-generator/tree/main#specifications) as our existing builders.
```

To demonstrate the flexibility of this framework, we are also announcing three SLSA builders created by [community contributors](https://github.com/laurentsimon/slsa-github-generator/blob/feat/hof/README.md#builder-creation) for the Java ecosystems.

## Build Your Own Builder Framework

The BYOB framework benefits both GitHub Action maintainers and GitHub Action users:

> 1. For Action maintainers, it makes it easy to meet the [SLSA Build L3](https://slsa.dev/spec/v1.0/levels#build-l3).

Reviewer 2: ```suggestion
1. For Action maintainers, it makes it easy to meet the [SLSA Build L3](https://slsa.dev/spec/v1.0/levels#build-l3) requirements.
```

1. For Action users, it makes it easy to adopt SLSA by trusting the BYOB project and the Action code - without worrying about which machine runs the Action.

> The BYOB framework provides a set of GitHub Actions and workflows that helps builder writers generate provenance.

Reviewer 2: ```suggestion
The BYOB framework provides a set of GitHub Actions and workflows that helps builder authors generate provenance.
```

![Maven provenance link](https://github.com/slsa-framework/slsa/assets/64505099/14ff1de1-a30c-4683-860c-352dc490c1ef)

Once you have the link to the artifact, you can verify it as follows:

```shell
$ ARTIFACT=test-artifact-0.1.0-jar
$ ARTIFACT_URL="https://repo1.maven.org/maven2/path/to/${ARTFACT}"
$ PROVENANCE_URL="${ARTIFACT_URL}".build.slsa
$ wget "${ARTIFACT_URL}" && wget "${PROVENANCE_URL}"
$ slsa-verifier verify-artifact "${ARTIFACT}" --provenance-path="${ARTIFACT}.build.slsa" --source-uri github.com/org/repo
```

### Verification for your dependencies

A Java project contains not only the main application code, but also its dependencies consumed as "packages". We can therefore recursively verify each dependency's provenance. To this end, we're releasing an [experimental Maven plugin](https://github.com/slsa-framework/slsa-verifier/tree/main/experimental/maven-plugin). The plugin resolves all dependencies of a given project and checks if they have provenance attestations along with their releases. When the plugin finds a dependency that has a provenance statement, it verifies it against the dependency.

> The plugin automatically performs the verification when configured in the ["pom.xml"](TODO:link to doc). Or it may be run [manually](TODO:link to doc). See our [documentation](https://github.com/slsa-framework/slsa-verifier/tree/main/experimental/maven-plugin).

Reviewer 2: ```suggestion

The plugin automatically performs the verification when configured in the ["pom.xml"](TODO:link to doc). Or it may be run [manually](TODO:link to doc). See our [documentation](https://github.com/slsa-framework/slsa-verifier/tree/main/experimental/maven-plugin).
```

---
title: "Build your own SLSA 3+ provenance builder on GitHub Actions"
author: "Andres Almiray (JReleaser), Adam Korczynski (Ada Logics), Philip Harrison (GitHub), Laurent Simon (Google)"
is_guest_post: false
---

It has been an exciting quarter for supply chain security and SLSA, with the release of the [SLSA v1.0 specification](2023-04-19-slsa-v1-final.md), [SLSA provenance support for npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/), and the announcement of new SLSA Level 3 builders for [Node.js](2023-05-11-bringing-improved-supply-chain-security-to-the-nodejs-ecosystem.md) and [containers](2023-06-13-slsa-github-worfklows-container-based.md)!

SLSA now provides and maintains official builders for [Go](2022-06-20-slsa-github-workflows.md), [Node.js](2023-05-11-bringing-improved-supply-chain-security-to-the-nodejs-ecosystem.md) and [Container](2023-06-13-slsa-github-worfklows-container-based.md) based projects. But what if you don't use any of these languages or use custom tooling that isn't supported by the official builders?

> To empower the community to create their own provenance builders and leverage the secure architecture of the official SLSA builders we are releasing the ["Build Your Own Builder" (BYOB) framework](https://github.com/slsa-framework/slsa-github-generator/tree/main#build-your-own-builder) for GitHub Actions. This makes it easy to take an existing GitHub Action (e.g. [JReleaser](https://jreleaser.org/)) and make it produce [SLSA Build Level 3 provenance](/spec/v1.0/requirements#provenance-generation).

Reviewer 3: Suggestion: Have a more direct answer to the question in the previous sentence/paragraph.

Reviewer 3: It could for example be:

change 

"To empower the community to create their own provenance builders and leverage the secure architecture of the official SLSA builders we are releasing the ["Build Your Own Builder" (BYOB) framework](https://github.com/slsa-framework/slsa-github-generator/tree/main#build-your-own-builder) for GitHub Actions."

to

"Today we release ["Build Your Own Builder" (BYOB) framework](https://github.com/slsa-framework/slsa-github-generator/tree/main#build-your-own-builder) for GitHub Actions which will empower the community to create their own provenance builders and leverage the secure architecture of the official SLSA builders."

To demonstrate the flexibility of this framework, we are also announcing three SLSA builders created by [community contributors](https://github.com/laurentsimon/slsa-github-generator/blob/feat/hof/README.md#builder-creation) for the Java ecosystems.

## Build Your Own Builder Framework

The BYOB framework benefits both GitHub Action maintainers and GitHub Action users:

1.  For Action maintainers, it makes it easy to meet the [SLSA Build L3](/spec/v1.0/levels#build-l3) requirements.
2.  For Action users, it makes it easy to adopt SLSA by trusting the BYOB project and the Action code - without worrying about which machine runs the Action.

The BYOB framework provides a set of GitHub Actions and workflows that helps builder authors generate provenance. Suppose you own a GitHub Action called `MyAction` and want to generate provenance showing that it ran on some input and generated some output, without having to trust the Workflow that called your Action. This is not possible using a regular Action because Actions run under the control of the calling Workflow (the diagram below depicts a project's `release.yml` workflow calling `MyAction`).

> ![action-release](https://github.com/slsa-framework/slsa/assets/64505099/367ecc46-28f6-4029-853e-161a028e6a35)

Author: Trying to address Mark's earlier comment to show that MyAction runs on the same machine as the workflow

> To solve this problem, you could turn your Action into a Reusable Workflow. This results in `MyAction` running in a VM under your control, not the caller's control. In fact, this is how the SLSA go, Node.js, and container builders work. This option is depicted in the diagram below: The project's `release.yml` calls the Reusable Workflow `MyReusableWorkflow` which in turn calls `MyAction` and generates provenance for the run.

Author: great catch. Thank you!

> To solve this problem, you could turn your Action into a Reusable Workflow. This results in `MyAction` running in a VM under your control, not the caller's control. In fact, this is how the SLSA go, Node.js, and container builders work. This option is depicted in the diagram below: The project's `release.yml` calls the Reusable Workflow `MyReusableWorkflow` which in turn calls `MyAction` and generates provenance for the run.

Author: sorry, this option does work. The Action is _not_ under the caller's control. SO it works, but it's tedious to implement and maintain all this. How shall I rephrase to make this more obvious?

> To solve this problem, you could turn your Action into a Reusable Workflow. This results in `MyAction` running in a VM under your control, not the caller's control. In fact, this is how the SLSA go, Node.js, and container builders work. This option is depicted in the diagram below: The project's `release.yml` calls the Reusable Workflow `MyReusableWorkflow` which in turn calls `MyAction` and generates provenance for the run.

Reviewer 4: I'm sure it works 🙂 What I mean is: what is the threat model, and the problems you solved? Right now, it simply isn't very clear. This can be easily fixed.

> To solve this problem, you could turn your Action into a Reusable Workflow. This results in `MyAction` running in a VM under your control, not the caller's control. In fact, this is how the SLSA go, Node.js, and container builders work. This option is depicted in the diagram below: The project's `release.yml` calls the Reusable Workflow `MyReusableWorkflow` which in turn calls `MyAction` and generates provenance for the run.

Author: Can you make a suggestion? I'm running out of ideas :/

![action-reusable](https://github.com/slsa-framework/slsa/assets/64505099/a0603e5f-4ebb-4c93-8216-b63f22bcf08d)

However, this is a lot of work that requires careful design and implementation. That's where the BYOB framework comes in! BYOB offloads all of the security critical work so that you can wrap your Action in a Reusable Workflow and call BYOB to do the heavy lifting. This high-level architecture is depicted in the diagram below: The `MyReusableWorkflow` calls the `BYOBWorkflow` which acts as an orchestrator.

![BYOB architecture](https://github.com/slsa-framework/slsa/assets/64505099/9d0a8133-ae1a-4b43-b7a1-5090e263eb47)

> There are two main steps to using the BYOB framework. First, the builder (`MyReusableWorkflow`) initializes the BYOB framework (`BYOB_Initialize` box in the middle box). Then it calls the framework (`BYOB_Run` box). Running the framework transfers execution to the BYOB framework which will run the `MyAction` in an isolated environment and then generate provenance.

Author: where would you like to move it to?

3.  A [Gradle builder](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/gradle#readme), contributed by [Ada Logics](https://adalogics.com). This builder is currently hosted in the OpenSSF SLSA repository.

These Java builders can publish provenance attestation on Maven central. Additionally, the JReleaser SLSA builder can provide attestation for artifacts published as GitHub release assets and/or uploaded to cloud storage such as AWS S3. Like other builders released by the SLSA Tooling SIG, the provenance can be verified using the [slsa-verifier](https://github.com/slsa-framework/slsa-verifier).

### Verification of Artifact

To verify the provenance of an artifact built by the Maven builder, we need to download the artifact and its provenance. In this example, we download them from Maven central. Maven Central lets users browse the files of each release, and from there we can find the artifacts along with their provenance files:

![Maven provenance link](https://github.com/slsa-framework/slsa/assets/64505099/14ff1de1-a30c-4683-860c-352dc490c1ef)

We download and verify it as follows:

> ```shell

Reviewer 4: This is not very user-friendly, but I suppose it works for now. Also, please concretely state exactly what security issues this verification does (and does not yet) solve for the unfamiliar reader.

Author: added a sentence, ptal

---
title: "Build your own SLSA 3+ provenance builder on GitHub Actions"
author: "Andres Almiray (JReleaser), Adam Korczynski (AdaLogics), Philip Harrison (GitHub), Laurent Simon (Google)"
is_guest_post: false
---

> It has been an exciting quarter for supply chain security and SLSA, with the release of the [SLSA v1.0 specification](https://slsa.dev/blog/2023/04/slsa-v1-final), [SLSA provenance support for npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/), and the announcement of new SLSA Level 3 builders for [Node.js](https://slsa.dev/blog/2023/05/bringing-improved-supply-chain-security-to-the-nodejs-ecosystem) and [containers](https://slsa.dev/blog/2023/06/slsa-github-worfklows-container-based)! 

Reviewer 5: suggestion: It would be helpful to explain what the main benefit of BYOB is (i.e. what the alternative to BYOB is) and how it differs from the generic SLSA builder. Right now it seems like, "You can do X" but I'm left wondering why.

Maybe something like:

[first paragraph: stating what we've done before, basically what you have now]

[second paragraph: setting up the problem and why existing solutions aren't suitable - expand on "what if you don't use any of these"]

[third paragraph: announcing BYOB and stating why it solves the problems above - combine "to empower" and "as a tool", but also go into more detail about *why* it's better]

[fourth paragraph: lead into the rest of the post]

Author: @olivekl can you chime in?

Reviewer 6: +1 to adding the why. I like Mark's suggested structure, or if you're trying to make fewer changes, you could just add a few sentences at the beginning of the first body section ("BYOB Framework") that cover these points in one sentence each.

Author: Thanks. I'll update this comment once I have made some changes

Author: @olivekl can you make a suggestion to add just a few sentences?

Author: I added the following paragraph:
```
Writing a builder from scratch is a tedious multi-month effort. The BYOB framework streamlines this process by cutting the development time down to a few days. As a builder writer, you need not worry about keeping signing keys secure, isolation between builds, the creation of attestations; all this is handled seamlessly by the framework, using the [same security design principles](https://github.com/slsa-framework/slsa-github-generator/tree/main#specifications) as our existing builders.
```

Reviewer 5: Yes, and i made another suggestion along these lines. I'll mark this one resolved

> It has been an exciting quarter for supply chain security and SLSA, with the release of the [SLSA v1.0 specification](https://slsa.dev/blog/2023/04/slsa-v1-final), [SLSA provenance support for npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/), and the announcement of new SLSA Level 3 builders for [Node.js](https://slsa.dev/blog/2023/05/bringing-improved-supply-chain-security-to-the-nodejs-ecosystem) and [containers](https://slsa.dev/blog/2023/06/slsa-github-worfklows-container-based)! 

Author: I'm not following. Can you give an example?

> It has been an exciting quarter for supply chain security and SLSA, with the release of the [SLSA v1.0 specification](https://slsa.dev/blog/2023/04/slsa-v1-final), [SLSA provenance support for npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/), and the announcement of new SLSA Level 3 builders for [Node.js](https://slsa.dev/blog/2023/05/bringing-improved-supply-chain-security-to-the-nodejs-ecosystem) and [containers](https://slsa.dev/blog/2023/06/slsa-github-worfklows-container-based)! 

Reviewer 5: Any links starting with "https://slsa.dev" should instead be relative URLs. Furthermore, links to other files in the same directory are often easier to link to the `.md` file rather than the URL in the browser.

Example: https://github.com/slsa-framework/slsa/blob/main/docs/_posts/2023-04-19-slsa-v1-final.md?plain=1#L25

```suggestion
It has been an exciting quarter for supply chain security and SLSA, with the release of the [SLSA v1.0 specification](2023-04-19-slsa-v1-final.md), [SLSA provenance support for npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/), and the announcement of new SLSA Level 3 builders for [Node.js](2023-05-11-bringing-improved-supply-chain-security-to-the-nodejs-ecosystem.md) and [containers](2023-06-13-slsa-github-worfklows-container-based.md)! 
```

As a tool or ecosystem maintainer you can focus on enabling your ecosystem to produce SLSA Level 3 provenance attestations without needing to worry about keeping signing keys secure, isolation between builds, and the creation of attestations; all this is handled seamlessly by the framework, using the [same security design principles](https://github.com/slsa-framework/slsa-github-generator/tree/main#specifications) as our existing builders.

To demonstrate the flexibility of this framework, we are also announcing three SLSA builders created by [community contributors](https://github.com/laurentsimon/slsa-github-generator/blob/feat/hof/README.md#builder-creation) for the Java ecosystems.

> # Build Your Own Builder Framework

Reviewer 5: nit: All of these headings need to move down a level (`#` -> `##`, etc.).

Author: thanks, done

Reviewer 5: Not done? This is still an h1 (`#`) but every thing should be h2 (`##`) and below

> The BYOB framework provides a set of GitHub Actions and workflows that a builder calls to generate provenance. The high-level architecture is depicted in the diagram below. The `builder_workflow.yml` represents the builder being created. The BYOB framework, on the right of the diagram, acts as a dispatcher. 

Reviewer 5: Right now it's just "how to use" but doesn't explain how it works. That might be worth getting into. Otherwise, as mentioned above, I'm left wondering what this actually does under the hood.

Author: The BYOB.md is user documentation. We have a design doc that can be turned into a BYOB_TECHNICAL.md. 
Which questions come to mind that you'd like to see answered?

Reviewer 5: To reiterate, I suggest two things:

(1) Clarify who the audience ("you") is. At first I was assuming the reader was an end user (i.e. a developer who uses GitHub Actions to build), since the previous blog posts were focused on end users and the text says "what if you don’t use any of these languages?". But in reality the audience is the owner of an Action. It took me a while to realize that. Stating that up front would help.

(2) Clarify the what BYOB is and what the benefits are. In short, this wraps an Action around a reusable workflow such that consumers have confidence that the Action actually ran.
- For the Action owner, it makes it easier to get it to SLSA Build L3.
- For the Consumer, it makes it so that you only have to trust the BYOB infrastructure + the Action code (but not whatever machine runs the Action).

Author: I've made some updates. ptal

Reviewer 5: There are still some "yous" but I'll add more specific comments.

> ![BYOB architecture](https://github.com/slsa-framework/slsa/assets/64505099/51c796e3-754e-4cd8-b2ea-dd8c23662411)

Reviewer 7: I think the images should be checked into the repo somewhere rather than using the URL from the upload to the PR description.

Not sure where they should be located though.

Author: please someone advise. I don't know either. I looked at the other posts and had the same-looking URIs and I could not infer where images were stored

Reviewer 5: Does it work to include it in the _posts directory?

If not, let's just leave it as-is for this PR. Other posts use this same convention, so we can fix them all in one go.

There are two main steps to using the BYOB framework. First, you initialize the BYOB framework ("Initialize" box on the left). Then you call the framework ("Run" box on the left). Running the framework transfers execution to the BYOB framework which will run the build in an isolated environment and then generate provenance.

Let's see each of these steps in more detail.

> The snippet below shows how the initialization step is performed: the builder initializes the BYOB framework for the ubuntu-latest runner, with a build Action path "./internal/callback_action" and asks it to attest to its inputs. At runtime, the BYOB framework will isolate the "./internal/callback_action" into an ephemeral VM and run it on an ubuntu-latest runner. The call below returns a so-called "slsa-token" object which can then be used to run the framework itself.

Reviewer 5: Why is the initialization step needed? Can't it be folded into the one call? It would help to explain that to the reader who, like me, might be wondering.

Author: it's a quirk of how OIDC works on GitHub. We need to establish the identity of the caller, which can only be done via some code that runs in the context of the caller. I can try adding a line on it, but it's a little technical. Maybe better link to a design doc as a BYOB_TECHNICAL.md on the website?

Reviewer 5: So are you saying that the parameters to pass to the SLSA Reusable Workflow are complicated, so instead of having the user copy-paste that all, you encapsulate that in an Action that runs in the client (Tool Reusable Workflow)? If so, saying that directly might help.

Also this isn't explained at all in BYOB.md, and the [diagram](https://github.com/slsa-framework/slsa-github-generator/blob/main/BYOB.md#design-overview) implies that `setup-token` is another workflow rather than an action. That would be good to document.

Author: > Also this isn't explained at all in BYOB.md, and the [diagram](https://github.com/slsa-framework/slsa-github-generator/blob/main/BYOB.md#design-overview) implies that setup-token is another workflow rather than an action. That would be good to document.

I think we can improve the image. https://github.com/slsa-framework/slsa-github-generator/blob/main/BYOB.md#slsa-setup-action-ssa says it's an Action, but the diagram could be improved. +1

Reviewer 2: > but the BYOB framework needs a way to identify who's calling it, so that it can attest to it and its parameters

Curious about improvements to actions here. Would we not need the init stage if the BYOB framework could ask for an OIDC token that was minted in the caller context? Or are there other bits we need?

Author: good question. We may still want the separate init stage because reusable workflows can be chained. RW1 -> RW2 -> RW3 -> BYOB. If we're attesting to RW1, we want to ensure that RW2 and RW3 cannot tamper with the parameters. If that information was present in the GitHub token or an API, it would not be necessary.

Reviewer 7: Right, the key thing is we need the whole call chain of `job_workflow_ref`s and `inputs` or at least RW1s info to be in the token. We could explicitly only support one level of reusable workflow calls but we wouldn't be able to easily detect it if a builder workflow author did use multiple levels of RWs and that would result in it not working or potentially being insecure wrt `inputs` and `job_workflow_ref`.

Author: Resolving.

2. A [Maven builder](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/maven#readme), contributed by [AdaLogics](https://adalogics.com). This repository is currently hosted in the OpenSSF SLSA repository.

3. A [Gradle builder](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/gradle#readme), contributed by [AdaLogics](https://adalogics.com). This repository is currently hosted in the OpenSSF SLSA repository.

These Java builders can publish provenance attestation on Maven central. Additionally, the JReleaser SLSA builder can provide attestation for artifacts published as GitHub release assets and/or uploaded to cloud storage such as AWS S3. Like other builders released by the SLSA Tooling SIG, the provenance can be verified using the [slsa-verifier](https://github.com/slsa-framework/slsa-verifier).

> ## Verification for your projects

Reviewer 5: Who is the "you" here? Is this a user of Maven? A consumer of a Maven package? The owner of the Maven builder?

Author: Renamed to "Verification of Artifact"

Below is an example for verifying provenance generated by the Maven builder: You will need to know the link of the artifact from the platform it has been distributed to. For Java artifacts, this will typically be Maven central, but you can also verify artifacts released on other platforms like Github. 

Maven Central allows you to browse the files of each release, and from there you can find the artifacts along with their provenance files:

![Maven provenance link](https://github.com/slsa-framework/slsa/assets/64505099/14ff1de1-a30c-4683-860c-352dc490c1ef)

Once you have the link to the artifact, you can verify it as follows:

```shell

> $ export ARTIFACT=test-artifact-0.1.0-jar
> $ export ARTIFACT_URL="https://repo1.maven.org/maven2/path/to/${ARTFACT}"
> $ export PROVENANCE_URL="${ARTIFACT_URL}".build.slsa

Reviewer 5: nit: No need to export these variables.

```suggestion
$ ARTIFACT=test-artifact-0.1.0-jar
$ ARTIFACT_URL="https://repo1.maven.org/maven2/path/to/${ARTFACT}"
$ PROVENANCE_URL="${ARTIFACT_URL}".build.slsa
```

> $ export PROVENANCE_URL="${ARTIFACT_URL}".build.slsa

Reviewer 5: Not specific to this blog post, but it's unfortunate that they chose to use a non-standard extension. We should work within the Securing Repos WG to align on a convention, e.g. `.intoto.jsonl`. @di FYI.

Author: "they" is actually "we". We decided to use this extension. The builders on the repo can generate signed provenance in various formats. It used to be an intoto DSSE, now it's a sigstore bundle. We don't want to break users who rely on file extension when the underlying format changes, so we decided to use a "universal" extension that says "it's a SLSA provenance for the build track". 

A verifier needs to know the builder to verify its generated provenance, therefore it must be aware of the format it creates.

$ wget "${ARTIFACT_URL}" && wget "${PROVENANCE_URL}"
$ slsa-verifier verify-artifact "${ARTIFACT}" --provenance-path="${ARTIFACT}.build.slsa" --source-uri github.com/org/repo
```

> ## Verification for your dependencies

Reviewer 5: I don't really understand what this section is saying and who "you" is.

Author: Renamed to "Verification of Dependencies". Changed "you" to "we" in the text

A Java project contains not only the main application code, but also its dependencies consumed as "packages". We can therefore recursively verify each dependency's provenance. To this end, we're releasing an [experimental Maven plugin](https://github.com/slsa-framework/slsa-verifier/tree/main/experimental/maven-plugin). The plugin resolves all dependencies of a given project and checks if they have provenance attestations along with their releases. When the plugin finds a dependency that has a provenance statement, it verifies it against the dependency.

> The plugin automatically performs the verification when configured in the ["pom.xml"](TODO:link to doc). Or it may be run [manually](TODO:link to doc). See our [documentation](https://github.com/slsa-framework/slsa-verifier/tree/main/experimental/maven-plugin).

Reviewer 5: Reminder to add these links

Author: Links added. Resolving.

> The plugin is still POC, but it demonstrates what's possible with and likely to come from the BYOB Framework. With the framework producing L3 provenance for GitHub Actions users, and the plugin automatically verifying Maven packages, you can build an end-to-end solution for an entire ecosystem.

Reviewer 5: ```suggestion
The plugin is still a proof-of-concept, but it demonstrates what's possible with and likely to come from the BYOB Framework. With the framework producing L3 provenance for GitHub Actions users, and the plugin automatically verifying Maven packages, you can build an end-to-end solution for an entire ecosystem.
```

> ## Conclusion

Reviewer 5: ```suggestion
## Conclusion

```

---
title: "Build your own SLSA 3+ provenance builder on GitHub Actions"
author: "Andres Almiray (JReleaser), Adam Korczynski (AdaLogics), Philip Harrison (GitHub), Laurent Simon (Google)"
is_guest_post: false
---

It has been an exciting quarter for supply chain security and SLSA, with the release of the [SLSA v1.0 specification](2023-04-19-slsa-v1-final.md), [SLSA provenance support for npm](https://github.blog/2023-04-19-introducing-npm-package-provenance/), and the announcement of new SLSA Level 3 builders for [Node.js](2023-05-11-bringing-improved-supply-chain-security-to-the-nodejs-ecosystem.md) and [containers](2023-06-13-slsa-github-worfklows-container-based.md)! 

> SLSA now provides and maintains official builders for [Go](https://slsa.dev/blog/2022/06/slsa-github-workflows), [Node.js](https://slsa.dev/blog/2023/05/bringing-improved-supply-chain-security-to-the-nodejs-ecosystem) and [Container](https://slsa.dev/blog/2023/06/slsa-github-worfklows-container-based) based projects, but what if you don't use any of these languages or use custom tooling that isn't supported by the official builders?

Author: removed all the links.

To demonstrate the flexibility of this framework, we are also announcing three SLSA builders created by [community contributors](https://github.com/laurentsimon/slsa-github-generator/blob/feat/hof/README.md#builder-creation) for the Java ecosystems.

## Build Your Own Builder Framework

The BYOB framework benefits both GitHub Action maintainers and GitHub Action users:

1. For Action maintainers, it makes it easy to meet the [SLSA Build L3](https://slsa.dev/spec/v1.0/levels#build-l3) requirements.
1. For Action users, it makes it easy to adopt SLSA by trusting the BYOB project and the Action code - without worrying about which machine runs the Action.

The BYOB framework provides a set of GitHub Actions and workflows that helps builder authors generate provenance.

> The high-level architecture is depicted in the diagram below. The `builder_workflow.yml` represents the builder being created. The BYOB framework, on the right of the diagram, acts as an orchestrator. 

Reviewer 5: IMO this needs more introduction. Right now I'm not sure a reader will understand what `builder_workflow.yml` is, what the delta is before and after, or what is being accomplished. I suggest something like the following:

---8<---

Suppose you own a GitHub Action called `MyCoolAction` and want to generate provenance showing that it ran on some input and generated some output, without having to trust whatever Workflow called your Action. This is not possible using a regular Action because Actions run under the control of the calling Workflow.

(diagram showing `MyCoolAction` running within each of the callers' workflows)

To solve this problem, you could turn your Action into a Reusable Workflow. This results in `MyCoolAction` running in a VM under your control, not the caller's control. In fact, this is how the SLSA go, Node.js, and container builders work. (Q: Does this require changes to how clients call `MyCoolAction`? If so that should be documented.)

(diagram showing callers calling a central `MyCoolAction` reusable workflow, which runs the Action and generates provenance)

However, this is a lot of work can requires careful design and implementation. That's where the BYOB framework comes in! BYOB offloads all of the security critical work so that you can wrap your Action in a Reusable Workflow and call BYOB to do the heavy lifting.

(diagram showing callers calling `MyCoolAction` reusable workflow, which in turn calls the BYOB, which runs the Action and generates provenance)

Author: thanks for the input. I've made the changes. PTAL
